pkg/virtualgarden/kube_api_server_secrets.go

```diff
@@ -0,0 +1,380 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package virtualgarden
+
+import (
+	"context"
+	cryptorand "crypto/rand"
+	_ "embed"
+	"strings"
+
+	secretsutil "github.com/gardener/gardener/pkg/utils/secrets"
+
+	"github.com/ghodss/yaml"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	configv1 "k8s.io/apiserver/pkg/apis/config/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/gardener/gardener/pkg/utils"
+	v1 "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+const (
+	KubeApiServerSecretNameAdmissionKubeconfig = Prefix + "-kube-apiserver-admission-kubeconfig"
+	KubeApiServerSecretNameAuditWebhookConfig  = "kube-apiserver-audit-webhook-config"
+	KubeApiServerSecretNameAuthWebhookConfig   = Prefix + "-kube-apiserver-auth-webhook-config"
+	KubeApiServerSecretNameBasicAuth           = Prefix + "-kube-apiserver-basic-auth"
+	KubeApiServerSecretNameEncryptionConfig    = Prefix + "-kube-apiserver-encryption-config"
+	KubeApiServerSecretNameServiceAccountKey   = Prefix + "-service-account-key"
+)
+
+//go:embed resources/validating-webhook-kubeconfig.yaml
+var validatingWebhookKubeconfig []byte
+
+//go:embed resources/mutating-webhook-kubeconfig.yaml
+var mutatingWebhookKubeconfig []byte
+
+func (o *operation) deployKubeAPIServerSecrets(ctx context.Context, checksums map[string]string) (string, error) {
+	o.log.Infof("Deploying secrets for the kube-apiserver")
+
+	if err := o.deployKubeApiServerSecretAdmissionKubeconfig(ctx); err != nil {
+		return "", err
+	}
+
+	if err := o.deployKubeApiServerSecretAuditWebhookConfig(ctx, checksums); err != nil {
+		return "", err
+	}
+
+	if err := o.deployKubeApiServerSecretAuthWebhookConfig(ctx, checksums); err != nil {
+		return "", err
+	}
+
+	basicAuthPw, err := o.deployKubeApiServerSecretBasicAuth(ctx, checksums)
+	if err != nil {
+		return "", err
+	}
+
+	if err := o.deployKubeApiServerSecretEncryptionConfig(ctx, checksums); err != nil {
+		return "", err
+	}
+
+	if err := o.deployKubeApiServerSecretServiceAccountKey(ctx, checksums); err != nil {
+		return "", err
+	}
+
+	return basicAuthPw, nil
+}
+
+func (o *operation) deleteKubeAPIServerSecrets(ctx context.Context) error {
+	o.log.Infof("Deleting secrets for the kube-apiserver")
+
+	for _, name := range []string{
+		KubeApiServerSecretNameAdmissionKubeconfig,
+		KubeApiServerSecretNameAuditWebhookConfig,
+		KubeApiServerSecretNameAuthWebhookConfig,
+		KubeApiServerSecretNameBasicAuth,
+		KubeApiServerSecretNameEncryptionConfig,
+		KubeApiServerSecretNameServiceAccountKey,
+	} {
+		secret := o.emptySecret(name)
+		if err := o.client.Delete(ctx, secret); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *operation) deployKubeApiServerSecretAdmissionKubeconfig(ctx context.Context) error {
+	if !o.isWebhookEnabled() {
+		return nil
+	}
```

Comment:
is the corresponding config map  `virtual-garden-kube-apiserver-admission-config` containing the `AdmissionConfiguration` deployed at a later step?

Reply:
Just wanting to make sure it is not forgotten.

Reply:
Yes. Another pull request will contain the deployment of the configmaps `virtual-garden-kube-apiserver-admission-config` and `kube-apiserver-audit-policy-config`.